News Digest – Fetch news workflow failing, `ValueError: year 0 is out of range`

The news digest hasn't included any news since Jan 19, as it seems to be choking on an article that can't be parsed as expected.

Parser error originating from:
https://github.com/responsible-ai-collaborative/nlp-monitoring/blame/ebef42d6ee21c986b6c11fcc9ab937fd7052b8e3/fetch_news.py#L130

Workflow log:
https://github.com/responsible-ai-collaborative/nlp-monitoring/actions/runs/7362458113/job/20040909916
```
 Fetching https://www.wired.com/story/congress-senate-tech-companies-pay-ai-training-data/
URL already processed. Skipping...
Traceback (most recent call last):
  File "/home/runner/.local/lib/python3.10/site-packages/dateutil/parser/_parser.py", line 649, in parse
    ret = self._build_naive(res, default)
  File "/home/runner/.local/lib/python3.10/site-packages/dateutil/parser/_parser.py", line 1235, in _build_naive
    naive = default.replace(**repl)
ValueError: year 0 is out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/runner/work/nlp-monitoring/nlp-monitoring/nlp-monitoring/fetch_news.py", line 335, in <module>
    if __name__ == "__main__": main()
  File "/home/runner/work/nlp-monitoring/nlp-monitoring/nlp-monitoring/fetch_news.py", line 63, in main
    delete_old_article_text(mongo_client)
  File "/home/runner/work/nlp-monitoring/nlp-monitoring/nlp-monitoring/fetch_news.py", line 130, in delete_old_article_text
    article_date = dateutil.parser.parse(
  File "/home/runner/.local/lib/python3.10/site-packages/dateutil/parser/_parser.py", line 1368, in parse
    return DEFAULTPARSER.parse(timestr, **kwargs)
  File "/home/runner/.local/lib/python3.10/site-packages/dateutil/parser/_parser.py", line 651, in parse
    six.raise_from(ParserError(str(e) + ": %s", timestr), e)
  File "<string>", line 3, in raise_from
dateutil.parser._parser.ParserError: year 0 is out of range: 0000-12-27
Error: Process completed with exit code 1.
```

--- a/fetch_news.py
+++ b/fetch_news.py
@@ -127,17 +127,31 @@ def get_mean_embedding(mongo_client = None, connection_string = None):
 def delete_old_article_text(mongo_client):
     candidates_collection = mongo_client['aiidprod'].candidates
     for article in candidates_collection.find({'text': {'$exists': True}}):
-        article_date = dateutil.parser.parse(
-            article.get('date_published') or 
-            article.get('date_scraped') or 
-            '2023-08-30' # Date at which date_scraped started being collected.
-        )
-        article_age = datetime.datetime.now() - article_date
-        if article.get('text') and article_age.days > 30:
-            candidates_collection.update_one(
-                { 'url': article['url'] },
-                {'$unset': {'text': '', 'plain_text': ''}}
-            )
+        try:
+            article_date = None
+            if article.get('date_published'):
+                try:
+                    article_date = dateutil.parser.parse(article['date_published'])
+                except:
+                    pass
+                if not article_date and article.get('date_scraped'):
+                    try:
+                        article_date = dateutil.parser.parse(article['date_scraped'])
+                    except:
+                        pass
+                if not article_date:
+                    # Date at which date_scraped started being collected.
+                    article_date = dateutil.parser.parse('2023-08-30') 
+                  
+                article_age = datetime.datetime.now() - article_date
+                if article.get('text') and article_age.days > 30:
+                    candidates_collection.update_one(
+                        { 'url': article['url'] },
+                        {'$unset': {'text': '', 'plain_text': ''}}
+                    )
+        except Exception as ex:
+            print("Could not delete article", article)
+            traceback.print_exception(type(ex), ex, ex.__traceback__)
 
 def process_url(
     article_url,

